File: src/services/arbitrage_bot.py
from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Protocol

from src.core.bot_config import BotConfig
from src.core.polymarket_client import PolymarketClientService
from src.strategies.arbitrage_pure.detector import ArbitragePureDetector
from src.strategies.arbitrage_pure.executor import ArbitragePureExecutor
from src.strategies.arbitrage_pure.types import TradeLog


class Logger(Protocol):
    def log_error(self, message: str) -> None: ...

    def log_info(self, message: str) -> None: ...


@dataclass
class ArbitrageBotService:
    poly_client: PolymarketClientService
    config: BotConfig
    logger: Logger

    detector: ArbitragePureDetector = field(init=False)
    executor: ArbitragePureExecutor = field(init=False)

    _running: bool = field(default=False, init=False)
    _balance: float = field(default=0.0, init=False)
    _trade_logs: List[TradeLog] = field(default_factory=list, init=False)
    _stop_event: asyncio.Event | None = field(default=None, init=False, repr=False)

    def __post_init__(self) -> None:
        self.detector = ArbitragePureDetector(
            self.poly_client,
            min_profit_threshold=self.config.min_profit_threshold,
            min_market_volume=self.config.min_market_volume,
        )
        self.executor = ArbitragePureExecutor(self.poly_client, self.logger)
        self._balance = float(self.config.starting_capital)

    async def start(self) -> None:
        self._running = True
        self._stop_event = asyncio.Event()

        await self.poly_client.initialize()

        self.logger.log_info(
            f"Starting pure arbitrage bot | balance=${self._balance:.2f} paper={self.config.paper_trading} mock={self.config.mock_mode}"
        )

        while self._running:
            started = time.monotonic()
            try:
                await self.scan_loop()
            except Exception as e:
                self.logger.log_error(f"scan_loop error: {e}")
            elapsed = time.monotonic() - started

            sleep_for = max(0.0, (self.config.scan_interval_ms / 1000.0) - elapsed)

            # Allow stop() to interrupt the sleep.
            try:
                if self._stop_event is not None:
                    await asyncio.wait_for(self._stop_event.wait(), timeout=sleep_for)
                else:
                    await asyncio.sleep(sleep_for)
            except asyncio.TimeoutError:
                pass

    async def scan_loop(self) -> None:
        if not self._running:
            return

        min_amount = self._position_amount()
        if self._balance < min_amount or min_amount <= 0:
            self.logger.log_info(
                f"Balance too low to trade | balance=${self._balance:.2f} position_size=${self.config.position_size:.2f}"
            )
            return

        scan_limit = self.config.market_scan_limit
        self.logger.log_info(f"🔍 Scanning {scan_limit} markets...")

        opportunities = await self.detector.scan_markets(limit=scan_limit)
        self.logger.log_info(f"🎯 Found {len(opportunities)} opportunities")

        for arb in opportunities:
            if not self._running:
                break

            amount = self._position_amount()
            if amount <= 0:
                break

            self.logger.log_info(
                f"⚡ Executing: {arb.question or arb.market} ({arb.profit * 100:.2f}% expected profit)"
            )

            op_started = time.monotonic()
            result = await self.executor.execute(arb, amount=amount)
            op_time = time.monotonic() - op_started

            if result.success:
                self._update_balance(result.profit)
                self._trade_logs.append(
                    TradeLog(
                        timestamp=datetime.now(),
                        market=result.market,
                        type=result.type,
                        profit=result.profit,
                        balance=self._balance,
                        operation_time=op_time,
                    )
                )
                self.logger.log_info(
                    f"✅ COMPLETED: +${result.profit:.2f} profit | Balance: ${self._balance:.2f}"
                )
            else:
                self.logger.log_error(f"❌ FAILED: {result.error}")

            await asyncio.sleep(self.config.execution_delay_ms / 1000.0)

    def stop(self) -> None:
        self._running = False
        if self._stop_event is not None:
            self._stop_event.set()
        self.logger.log_info("Stopping bot...")

    def get_balance(self) -> float:
        return float(self._balance)

    def get_trade_logs(self) -> List[TradeLog]:
        return list(self._trade_logs)

    def _position_amount(self) -> float:
        if self._balance <= 0:
            return 0.0

        size = float(self.config.position_size)
        # Allow POSITION_SIZE<=1 to mean fraction of capital.
        if 0 < size <= 1:
            size = self._balance * size

        return min(size, self._balance)

    def _update_balance(self, profit: float) -> None:
        self._balance += float(profit)

File: src/services/arbitrage-bot.ts
import { BotConfig } from '../config/bot-config.js';
import { PolymarketClientService } from '../core/polymarket-client.js';
import { ArbitragePureDetector, ArbitragePureExecutor, TradeLog } from '../strategies/arbitrage-pure/index.js';

export interface Logger {
    logError(message: string): void;
    logInfo(message: string): void;
}

export class ArbitrageBotService {
    private detector: ArbitragePureDetector;
    private executor: ArbitragePureExecutor;
    private _running: boolean = false;
    private _balance: number = 0.0;
    private _tradeLogs: TradeLog[] = [];
    private _stopPromise: ((value: void) => void) | null = null;
    private cycleCount: number = 0;

    constructor(
        private polyClient: PolymarketClientService,
        private config: BotConfig,
        private logger: Logger
    ) {
        this.detector = new ArbitragePureDetector(
            this.polyClient,
            this.config.min_profit_threshold,
            this.config.min_market_volume
        );
        this.executor = new ArbitragePureExecutor(this.polyClient, this.logger);
        this._balance = this.config.starting_capital;
    }

    async start(): Promise<void> {
        this._running = true;

        this.logger.logInfo(
            `Starting pure arbitrage bot | balance=$${this._balance.toFixed(2)} paper=${this.config.paper_trading} mock=${this.config.mock_mode}`
        );

        while (this._running) {
            const started = Date.now();
            try {
                await this.scanLoop();
            } catch (e: any) {
                this.logger.logError(`scanLoop error: ${e.message}`);
            }
            const elapsed = Date.now() - started;
            const sleepFor = Math.max(0, this.config.scan_interval_ms - elapsed);

            if (sleepFor > 0 && this._running) {
                await new Promise<void>((resolve) => {
                    const timeout = setTimeout(resolve, sleepFor);
                    this._stopPromise = () => {
                        clearTimeout(timeout);
                        resolve();
                    };
                });
            }
        }
    }

    async scanLoop(): Promise<void> {
        if (!this._running) return;
        
        const cycleStartTime = Date.now();
        this.cycleCount++;
        
        console.log(`\n${'='.repeat(70)}`);
        console.log(`📍 CYCLE #${this.cycleCount} | ${new Date().toISOString()}`);
        console.log(`${'='.repeat(70)}`);

        try {
            // SCAN
            const scanStartTime = Date.now();
            console.log(`⏳ Starting scan...`);

            const scanLimit = this.config.market_scan_limit;
            const opportunities = await this.detector.scanMarkets(scanLimit);
            
            const scanTime = Date.now() - scanStartTime;
            console.log(`\n✅ Scan finished in ${(scanTime / 1000).toFixed(2)}s`);
            console.log(`🎯 Found ${opportunities.length} opportunities`);

            if (opportunities.length === 0) {
                console.log(`⏸️ No opportunities. Waiting for next cycle...\n`);
                return;
            }

            // EXECUTE
            console.log(`\n⚡ Executing ${opportunities.length} opportunities...`);
            
            let executedCount = 0;
            let totalProfit = 0;
            
            for (const arb of opportunities.slice(0, 5)) {
                if (!this._running) break;

                const amount = this.positionAmount();
                if (amount <= 0) {
                    console.log(`⚠️ Balance too low to continue executing.`);
                    break;
                }

                console.log(`\n  ⚡ [${executedCount + 1}/${Math.min(5, opportunities.length)}] ${arb.question || arb.market}`);
                
                const execStartTime = Date.now();
                const result = await this.executor.execute(arb, amount);
                const execTime = Date.now() - execStartTime;

                if (result.success) {
                    executedCount++;
                    totalProfit += result.profit;
                    this.updateBalance(result.profit);
                    
                    this._tradeLogs.push({
                        timestamp: new Date(),
                        market: result.market,
                        type: result.type,
                        profit: result.profit,
                        balance: this._balance,
                        operation_time: execTime / 1000,
                    });
                    
                    console.log(`  ✅ +$${result.profit.toFixed(2)} (${(execTime / 1000).toFixed(2)}s) | Balance: $${this._balance.toFixed(2)}`);
                } else {
                    console.log(`  ❌ Failed: ${result.error}`);
                }

                if (this.config.execution_delay_ms > 0) {
                    await new Promise((resolve) => setTimeout(resolve, this.config.execution_delay_ms));
                }
            }
            
            // SUMMARY
            const cycleTime = Date.now() - cycleStartTime;
            console.log(`\n${'='.repeat(70)}`);
            console.log(`📊 CYCLE #${this.cycleCount} SUMMARY`);
            console.log(`├─ Total time: ${(cycleTime / 1000).toFixed(2)}s`);
            console.log(`├─ Opportunities: ${opportunities.length}`);
            console.log(`├─ Executed: ${executedCount}`);
            console.log(`├─ Profit: +$${totalProfit.toFixed(2)}`);
            console.log(`├─ Balance: $${this._balance.toFixed(2)}`);
            console.log(`└─ Next cycle in 30 seconds...`);
            console.log(`${'='.repeat(70)}\n`);

        } catch (error) {
            console.log(`\n❌ Cycle error:`, error);
        }
    }

    stop(): void {
        this._running = false;
        if (this._stopPromise) {
            this._stopPromise();
        }
        this.logger.logInfo("Stopping bot...");
    }

    getBalance(): number {
        return this._balance;
    }

    getTradeLogs(): TradeLog[] {
        return [...this._tradeLogs];
    }

    private positionAmount(): number {
        if (this._balance <= 0) return 0;
        let size = this.config.position_size;
        if (size > 0 && size <= 1) {
            size = this._balance * size;
        }
        return Math.min(size, this._balance);
    }

    private updateBalance(profit: number): void {
        this._balance += profit;
    }
}
